src/Giraffe/HttpHandlers.fs

```diff
@@ -43,17 +45,34 @@ let private getPath (ctx : HttpContext) =
     | Some p -> ctx.Request.Path.ToString().[p.Length..]
     | None   -> ctx.Request.Path.ToString()
 
+// let private handlerWithRootedPath (path : string) (handler : HttpHandler) : HttpHandler =
+//     fun (next : HttpFunc) (ctx : HttpContext) ->
+//         task {
+//             let savedSubPath = getSavedSubPath ctx
+//             try
+//                 ctx.Items.Item RouteKey <- ((savedSubPath |> Option.defaultValue "") + path)
+//                 return! handler next ctx
+//             finally
+//                 printfn "saved subpath finalising is: %A on path %s" savedSubPath path 
+//                 match savedSubPath with
+//                 | Some savedSubPath -> ctx.Items.Item   RouteKey <- savedSubPath.Substring(0,savedSubPath.Length - path.Length)
+//                 | None              -> ctx.Items.Remove RouteKey |> ignore
+//         }
+
 let private handlerWithRootedPath (path : string) (handler : HttpHandler) : HttpHandler =
     fun (next : HttpFunc) (ctx : HttpContext) ->
-        async {
+        task {
             let savedSubPath = getSavedSubPath ctx
-            try
-                ctx.Items.Item RouteKey <- ((savedSubPath |> Option.defaultValue "") + path)
-                return! handler next ctx
-            finally
+            ctx.Items.Item RouteKey <- ((savedSubPath |> Option.defaultValue "") + path)
+            let rtask = handler next ctx
+            let! result = rtask
+            match result with
+            | Some _ -> () 
+            | None ->
                 match savedSubPath with
-                | Some savedSubPath -> ctx.Items.Item RouteKey <- savedSubPath
+                | Some savedSubPath -> ctx.Items.Item   RouteKey <- savedSubPath.Substring(0,savedSubPath.Length - path.Length)
                 | None              -> ctx.Items.Remove RouteKey |> ignore
+            return! rtask
         }
 
```

Comment:
Ok, one more question, I am not sure why on line 73 you had to run `ctx.Items.Item   RouteKey <- savedSubPath.Substring(0,savedSubPath.Length - path.Length)`. Shouldn't assigning `ctx.Items.Item   RouteKey <- savedSubPath` just be fine, as `savedSubPath` should still hold the same value from above?

I tried the following change and all tests still pass:

```fsharp
let private handlerWithRootedPath (path : string) (handler : HttpHandler) : HttpHandler =
    fun (next : HttpFunc) (ctx : HttpContext) ->
        task {
            let savedSubPath = getSavedSubPath ctx
            ctx.Items.Item RouteKey <- ((savedSubPath |> Option.defaultValue "") + path)
            let! result = handler next ctx
            match result with
            | Some _ -> ()
            | None ->
                match savedSubPath with
                | Some savedSubPath -> ctx.Items.Item   RouteKey <- savedSubPath
                | None              -> ctx.Items.Remove RouteKey |> ignore
            return result
        }
```

I think I am missing a critical test here where one of the two implementations should fail...